### ISSUE TITLE
LDAP Password not correct, Invalid Credentials

### ISSUE THREAD

**Issue by mkuendig** [goharbor/harbor#12699]
I have a very strange issue I can't solve. LDAP does work on one machine (harbor 2.0.0 on docker) and does NOT work with the exact same config on harbor 2.0.2 on Kubernetes.

- We have tested the user/password pair directly to LDAP and that is ok.

- We get an unknow user error if we choose another user that doesn't exist.

Which means the FW, the network, the certs all are fine. Just the password, which we have entered 5 times with copy/paste and manually typed and verified.

This is the error that we see on the harbor 2.0.2 on kubernetes box:

2020-08-10T18:23:06Z [ERROR] [/server/error/error.go:53]: {"errors":[{"code":"INTERNAL_SERVER_ERROR","message":"LDAP connect fail, error: LDAP Result Code 49 \"Invalid Credentials\": "}]}
/harbor/src/server/error/error.go:81, github.com/goharbor/harbor/src/server/error.apiError
/harbor/src/server/error/error.go:50, github.com/goharbor/harbor/src/server/error.SendError
/harbor/src/common/api/base.go:67, github.com/goharbor/harbor/src/common/api.(*BaseAPI).RenderError
/harbor/src/common/api/base.go:228, github.com/goharbor/harbor/src/common/api.(*BaseAPI).SendInternalServerError
/harbor/src/core/api/ldap.go:103, github.com/goharbor/harbor/src/core/api.(*LdapAPI).Ping
/usr/local/go/src/reflect/value.go:460, reflect.Value.call
/usr/local/go/src/reflect/value.go:321, reflect.Value.Call
/harbor/src/vendor/github.com/astaxie/beego/router.go:853, github.com/astaxie/beego.(*ControllerRegister).ServeHTTP
/harbor/src/server/middleware/transaction/transaction.go:62, github.com/goharbor/harbor/src/server/middleware/transaction.Middleware.func1.1
/harbor/src/lib/orm/orm.go:63, github.com/goharbor/harbor/src/lib/orm.WithTransaction.func1
/harbor/src/server/middleware/transaction/transaction.go:71, github.com/goharbor/harbor/src/server/middleware/transaction.Middleware.func1
/harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
/usr/local/go/src/net/http/server.go:2007, net/http.HandlerFunc.ServeHTTP
/harbor/src/server/middleware/notification/notification.go:31, github.com/goharbor/harbor/src/server/middleware/notification.Middleware.func1
/harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
/usr/local/go/src/net/http/server.go:2007, net/http.HandlerFunc.ServeHTTP
/harbor/src/server/middleware/orm/orm.go:53, github.com/goharbor/harbor/src/server/middleware/orm.MiddlewareWithConfig.func1
/harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
/usr/local/go/src/net/http/server.go:2007, net/http.HandlerFunc.ServeHTTP
/harbor/src/server/middleware/readonly/readonly.go:77, github.com/goharbor/harbor/src/server/middleware/readonly.MiddlewareWithConfig.func1
/harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
/usr/local/go/src/net/http/server.go:2007, net/http.HandlerFunc.ServeHTTP
/harbor/src/server/middleware/security/security.go:62, github.com/goharbor/harbor/src/server/middleware/security.Middleware.func1.1
/usr/local/go/src/net/http/server.go:2007, net/http.HandlerFunc.ServeHTTP
/harbor/src/server/middleware/csrf/csrf.go:55, github.com/goharbor/harbor/src/server/middleware/csrf.attach.func1
/usr/local/go/src/net/http/server.go:2007, net/http.HandlerFunc.ServeHTTP
/harbor/src/vendor/github.com/gorilla/csrf/csrf.go:297, github.com/gorilla/csrf.(*csrf).ServeHTTP
/harbor/src/server/middleware/csrf/csrf.go:75, github.com/goharbor/harbor/src/server/middleware/csrf.Middleware.func2
/harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
/usr/local/go/src/net/http/server.go:2007, net/http.HandlerFunc.ServeHTTP
/harbor/src/server/middleware/session/session.go:34, github.com/goharbor/harbor/src/server/middleware/session.Middleware.func1.1
/usr/local/go/src/net/http/server.go:2007, net/http.HandlerFunc.ServeHTTP
/harbor/src/server/middleware/log/log.go:33, github.com/goharbor/harbor/src/server/middleware/log.Middleware.func1
/harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
/usr/local/go/src/net/http/server.go:2007, net/http.HandlerFunc.ServeHTTP
/harbor/src/server/middleware/requestid/requestid.go:37, github.com/goharbor/harbor/src/server/middleware/requestid.Middleware.func1
/harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
/usr/local/go/src/net/http/server.go:2007, net/http.HandlerFunc.ServeHTTP
/usr/local/go/src/net/http/server.go:2802, net/http.serverHandler.ServeHTTP
/usr/local/go/src/net/http/server.go:1890, net/http.(*conn).serve
2020/08/10 18:23:06.616 [1;44m[D][0m [transaction.go:62]  |      127.0.0.1|[97;41m 500 [0m|  62.459337ms|   match|[97;46m POST    [0m /api/v2.0/ldap/ping   r:/api/v2.0/ldap/ping
2020/08/10 18:23:08.582 [1;44m[D][0m [transaction.go:62]  |      11.0.46.9|[97;42m 200 [0m|    785.187µs|   match|[97;44m GET     [0m /api/v2.0/ping   r:/api/v2.0/ping

and yes, I have also downgraded to harbor 2.0.0 on kubernetes but I see the same error. I have completely re-installed harbor, with also deleting the PV and PVC, so that is a completely fresh harbor instance. Bloody LDAP just doesn't connect.

Any help on this please? I have wasted now many, many hours without getting this to work.

this is the config that I get back from the API (anonymized):

{
  "auth_mode": {
    "value": "ldap_auth",
    "editable": true
  },
  "count_per_project": {
    "value": -1,
    "editable": true
  },
  "email_from": {
    "value": "admin \u003csample_admin@mydomain.com\u003e",
    "editable": true
  },
  "email_host": {
    "value": "smtp.mydomain.com",
    "editable": true
  },
  "email_identity": {
    "value": "",
    "editable": true
  },
  "email_insecure": {
    "value": false,
    "editable": true
  },
  "email_port": {
    "value": 25,
    "editable": true
  },
  "email_ssl": {
    "value": false,
    "editable": true
  },
  "email_username": {
    "value": "sample_admin@mydomain.com",
    "editable": true
  },
  "http_authproxy_endpoint": {
    "value": "",
    "editable": true
  },
  "http_authproxy_server_certificate": {
    "value": "",
    "editable": true
  },
  "http_authproxy_skip_search": {
    "value": false,
    "editable": true
  },
  "http_authproxy_tokenreview_endpoint": {
    "value": "",
    "editable": true
  },
  "http_authproxy_verify_cert": {
    "value": true,
    "editable": true
  },
  "ldap_base_dn": {
    "value": "ou=user,ou=CI,ou=abcRollen,o=abc,c=ch",
    "editable": true
  },
  "ldap_filter": {
    "value": "",
    "editable": true
  },
  "ldap_group_admin_dn": {
    "value": "",
    "editable": true
  },
  "ldap_group_attribute_name": {
    "value": "",
    "editable": true
  },
  "ldap_group_base_dn": {
    "value": "ou=CI,ou=abcRollen,o=abc,c=ch",
    "editable": true
  },
  "ldap_group_membership_attribute": {
    "value": "memberof",
    "editable": true
  },
  "ldap_group_search_filter": {
    "value": "(objectClass=groupOfUniqueNames)",
    "editable": true
  },
  "ldap_group_search_scope": {
    "value": 2,
    "editable": true
  },
  "ldap_scope": {
    "value": 2,
    "editable": true
  },
  "ldap_search_dn": {
    "value": "cn=NEXUS.User,ou=funktionaleUser,ou=CI,ou=abcRollen,o=abc,c=ch",
    "editable": true
  },
  "ldap_timeout": {
    "value": 5,
    "editable": true
  },
  "ldap_uid": {
    "value": "uid",
    "editable": true
  },
  "ldap_url": {
    "value": "ldaps://enterprisedirectory.prod.abc.com:636",
    "editable": true
  },
  "ldap_verify_cert": {
    "value": true,
    "editable": true
  },
  "notification_enable": {
    "value": true,
    "editable": true
  },
  "oidc_client_id": {
    "value": "",
    "editable": true
  },
  "oidc_endpoint": {
    "value": "",
    "editable": true
  },
  "oidc_groups_claim": {
    "value": "",
    "editable": true
  },
  "oidc_name": {
    "value": "",
    "editable": true
  },
  "oidc_scope": {
    "value": "",
    "editable": true
  },
  "oidc_verify_cert": {
    "value": true,
    "editable": true
  },
  "project_creation_restriction": {
    "value": "everyone",
    "editable": true
  },
  "quota_per_project_enable": {
    "value": true,
    "editable": true
  },
  "read_only": {
    "value": false,
    "editable": true
  },
  "robot_token_duration": {
    "value": 43200,
    "editable": true
  },
  "scan_all_policy": {
    "value": null,
    "editable": true
  },
  "self_registration": {
    "value": false,
    "editable": true
  },
  "storage_per_project": {
    "value": -1,
    "editable": true
  },
  "token_expiration": {
    "value": 30,
    "editable": true
  },
  "uaa_client_id": {
    "value": "",
    "editable": true
  },
  "uaa_client_secret": {
    "value": "",
    "editable": true
  },
  "uaa_endpoint": {
    "value": "",
    "editable": true
  },
  "uaa_verify_cert": {
    "value": false,
    "editable": true
  }
}

**Comment by mkuendig** [comment]
Any news here? @ywk253100 I see you tagged yourself. Any chance to get any tips?

**Comment by mkuendig** [comment]
We got some more hints here in the logs below. It seems that harbor-core runs as the user harbor, but the filesystem is owned by the user root. So most likely the update in the webui of the password is not persisted to disk as harbor-core can't write the password to disk. Certainly the password listed below is not the password we have entered in the webui for LDAP settings.

Does that help in any way to get this resolved?

harbor [ /harbor ]$ ls -al /etc/core/
total 28
drwxr-xr-x 4 root root   4096 2020-08-24 08:00 .
drwxr-xr-x 1 root root   4096 2020-08-24 08:00 ..
-rw-r--r-- 1 root harbor   72 2020-08-24 07:54 app.conf
drwxrwsrwt 3 root harbor  120 2020-08-24 07:54 ca
-rw-r--r-- 1 root harbor   11 2020-08-24 07:54 key
-rw-r--r-- 1 root harbor 3243 2020-08-24 07:54 private_key.pem
drwxrwsrwx 2 root harbor 4096 2020-08-24 07:54 token
harbor [ /harbor ]$ cat /etc/core/key
h&ArbOr)61?harbor [ /harbor ]$
harbor [ /harbor ]$
harbor [ /harbor ]$ ps -ef
UID            PID  PPID C STIME TTY          TIME CMD
harbor           1     0 0 00:48 ?        00:00:00 sh /harbor/entrypoint.sh
harbor          15     1 0 00:49 ?        00:00:03 harbor_core
harbor          31     0 0 18:30 pts/0    00:00:00 bash
harbor          44    31 0 20:02 pts/0    00:00:00 ps -ef
harbor [ /harbor ]$ whoami
harbor
harbor [ /harbor ]$
